File: WorkBranch/backend/service/agent_service/service/workspace_service.py
import os
import uuid
from typing import Optional, Dict, Set
from pathlib import Path


class WorkspaceService:
    """工作区服务：管理工作区的创建、路径验证与权限控制"""

    def __init__(self, base_dir: str = None):
        self._base_dir = base_dir or "workspaces"
        self._workspaces: Dict[str, dict] = {}
        os.makedirs(self._base_dir, exist_ok=True)

    @property
    def base_dir(self) -> str:
        """返回工作区基础目录的绝对路径"""
        return os.path.abspath(self._base_dir)

    def register(
        self,
        workspace_id: Optional[str] = None,
        session_id: Optional[str] = None
    ) -> str:
        """
        注册并创建工作区
        
        Args:
            workspace_id: 可选的工作区ID，不提供则自动生成
            session_id: 可选的会话ID，不提供则自动生成
            
        Returns:
            注册的工作区ID
        """
        if not workspace_id:
            workspace_id = str(uuid.uuid4())[:8]
        if not session_id:
            session_id = str(uuid.uuid4())[:8]

        if workspace_id in self._workspaces:
            existing = self._workspaces[workspace_id]
            if existing.get("session_id") == session_id:
                print(f"[Workspace] 工作区已存在: {workspace_id}")
                return workspace_id

        self._workspaces[workspace_id] = {
            "id": workspace_id,
            "session_id": session_id,
            "status": "active",
            "created_at": None
        }

        workspace_path = self.get_workspace_path(session_id, workspace_id)
        is_new = not os.path.exists(workspace_path)
        os.makedirs(workspace_path, exist_ok=True)
        
        if is_new:
            print(f"[Workspace] 工作区已创建: {workspace_id}")
        else:
            print(f"[Workspace] 工作区信息已载入: {workspace_id}")
        print(f"[Workspace] 会话ID: {session_id}")
        print(f"[Workspace] 路径: {workspace_path}")
        return workspace_id

    def get_workspace_path(self, session_id: str, workspace_id: str) -> str:
        """
        获取工作区的完整路径
        
        Args:
            session_id: 会话ID
            workspace_id: 工作区ID
            
        Returns:
            工作区的绝对路径
        """
        return os.path.join(self.base_dir, session_id, workspace_id)

    def get_workspace_info(self, workspace_id: str) -> Optional[dict]:
        """获取工作区信息"""
        return self._workspaces.get(workspace_id)

    def get_workspace_dir(self, workspace_id: str) -> Optional[str]:
        """
        获取工作区目录路径
        
        Args:
            workspace_id: 工作区ID
            
        Returns:
            工作区目录的绝对路径，不存在返回 None
        """
        info = self._workspaces.get(workspace_id)
        if not info:
            return None
        return self.get_workspace_path(info["session_id"], workspace_id)

    def exists(self, workspace_id: str) -> bool:
        """检查工作区是否已注册"""
        return workspace_id in self._workspaces

    def list_all(self) -> Set[str]:
        """列出所有工作区ID"""
        return set(self._workspaces.keys())

    def validate_path(self, workspace_id: str, target_path: str) -> tuple[bool, str]:
        """
        验证目标路径是否在工作区范围内
        
        Args:
            workspace_id: 工作区ID
            target_path: 目标文件/目录路径
            
        Returns:
            (是否合法, 规范化后的绝对路径或错误信息)
        """
        workspace_dir = self.get_workspace_dir(workspace_id)
        if not workspace_dir:
            return False, f"工作区不存在: {workspace_id}"

        try:
            abs_target = os.path.abspath(target_path)
            abs_workspace = os.path.abspath(workspace_dir)
            
            if not abs_target.startswith(abs_workspace + os.sep) and abs_target != abs_workspace:
                return False, f"路径越界: {target_path} 不在工作区 {workspace_dir} 范围内"
            
            return True, abs_target
        except Exception as e:
            return False, f"路径验证失败: {str(e)}"

    def is_path_allowed(self, workspace_id: str, target_path: str) -> bool:
        """
        快速检查路径是否被允许访问
        
        Args:
            workspace_id: 工作区ID
            target_path: 目标路径
            
        Returns:
            是否允许访问
        """
        allowed, _ = self.validate_path(workspace_id, target_path)
        return allowed

    def resolve_path(self, workspace_id: str, relative_path: str) -> tuple[bool, str]:
        """
        将相对路径解析为工作区内的绝对路径并验证
        
        Args:
            workspace_id: 工作区ID
            relative_path: 相对于工作区的路径
            
        Returns:
            (是否合法, 绝对路径或错误信息)
        """
        workspace_dir = self.get_workspace_dir(workspace_id)
        if not workspace_dir:
            return False, f"工作区不存在: {workspace_id}"

        if os.path.isabs(relative_path):
            return self.validate_path(workspace_id, relative_path)

        full_path = os.path.join(workspace_dir, relative_path)
        return self.validate_path(workspace_id, full_path)

    def delete_workspace(self, workspace_id: str) -> bool:
        """
        删除工作区（包括目录和注册信息）
        
        Args:
            workspace_id: 工作区ID
            
        Returns:
            是否删除成功
        """
        info = self._workspaces.get(workspace_id)
        if not info:
            return False

        workspace_path = self.get_workspace_path(info["session_id"], workspace_id)
        
        try:
            if os.path.exists(workspace_path):
                import shutil
                shutil.rmtree(workspace_path)
            
            del self._workspaces[workspace_id]
            print(f"[Workspace] 工作区已删除: {workspace_id}")
            return True
        except Exception as e:
            print(f"[Workspace] 删除工作区失败: {e}")
            return False

    def list_sessions(self) -> Dict[str, list]:
        """
        列出所有会话及其工作区
        
        Returns:
            {session_id: [workspace_id1, workspace_id2, ...]}
        """
        sessions: Dict[str, list] = {}
        for ws_id, info in self._workspaces.items():
            session_id = info.get("session_id", "unknown")
            if session_id not in sessions:
                sessions[session_id] = []
            sessions[session_id].append(ws_id)
        return sessions

File: packages/backend/src/service/agent-service/service/workspace-service.ts
import * as fs from 'fs';
import * as path from 'path';
import { v4 as uuidv4 } from 'uuid';
import { fileStorage } from '../../../data/file-storage';

interface WorkspaceInfo {
  id: string;
  session_id: string;
  status: string;
  created_at: string | null;
}

export interface FileInfo {
  name: string;
  path: string;
  absolutePath: string;
  size: number;
  type: string;
  modifiedAt: string;
  isDirectory: boolean;
}

export interface FileTreeNode {
  name: string;
  path: string;
  isDirectory: boolean;
  size?: number;
  type?: string;
  modifiedAt?: string;
  children?: FileTreeNode[];
}

const MAX_FILE_SIZE = 1024 * 1024 * 1024; // 1GB

function getMimeType(filePath: string): string {
  const ext = path.extname(filePath).toLowerCase();
  const mimeTypes: Record<string, string> = {
    '.txt': 'text/plain',
    '.md': 'text/markdown',
    '.json': 'application/json',
    '.js': 'application/javascript',
    '.ts': 'application/typescript',
    '.tsx': 'application/typescript',
    '.jsx': 'application/javascript',
    '.html': 'text/html',
    '.css': 'text/css',
    '.xml': 'application/xml',
    '.yaml': 'application/x-yaml',
    '.yml': 'application/x-yaml',
    '.csv': 'text/csv',
    '.pdf': 'application/pdf',
    '.doc': 'application/msword',
    '.docx': 'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
    '.xls': 'application/vnd.ms-excel',
    '.xlsx': 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
    '.ppt': 'application/vnd.ms-powerpoint',
    '.pptx': 'application/vnd.openxmlformats-officedocument.presentationml.presentation',
    '.zip': 'application/zip',
    '.tar': 'application/x-tar',
    '.gz': 'application/gzip',
    '.rar': 'application/vnd.rar',
    '.7z': 'application/x-7z-compressed',
    '.png': 'image/png',
    '.jpg': 'image/jpeg',
    '.jpeg': 'image/jpeg',
    '.gif': 'image/gif',
    '.svg': 'image/svg+xml',
    '.webp': 'image/webp',
    '.ico': 'image/x-icon',
    '.mp3': 'audio/mpeg',
    '.wav': 'audio/wav',
    '.mp4': 'video/mp4',
    '.webm': 'video/webm',
    '.avi': 'video/x-msvideo',
    '.py': 'text/x-python',
    '.java': 'text/x-java-source',
    '.c': 'text/x-c',
    '.cpp': 'text/x-c++src',
    '.h': 'text/x-c',
    '.hpp': 'text/x-c++hdr',
    '.rs': 'text/x-rust',
    '.go': 'text/x-go',
    '.rb': 'text/x-ruby',
    '.php': 'text/x-php',
    '.swift': 'text/x-swift',
    '.kt': 'text/x-kotlin',
    '.scala': 'text/x-scala',
    '.sh': 'application/x-sh',
    '.bash': 'application/x-sh',
    '.ps1': 'application/x-powershell',
    '.sql': 'application/x-sql',
    '.db': 'application/x-sqlite3',
    '.sqlite': 'application/x-sqlite3',
  };
  return mimeTypes[ext] || 'application/octet-stream';
}

function isHiddenFile(filename: string): boolean {
  return filename.startsWith('.');
}

function isBinaryFile(filePath: string): boolean {
  const ext = path.extname(filePath).toLowerCase();
  const textExtensions = [
    '.txt', '.md', '.json', '.js', '.ts', '.tsx', '.jsx', '.html', '.css', '.xml',
    '.yaml', '.yml', '.csv', '.py', '.java', '.c', '.cpp', '.h', '.hpp', '.rs',
    '.go', '.rb', '.php', '.swift', '.kt', '.scala', '.sh', '.bash', '.ps1', '.sql',
    '.log', '.ini', '.cfg', '.conf', '.env', '.gitignore', '.dockerignore',
    '.editorconfig', '.eslintrc', '.prettierrc', '.babelrc', '.tsconfig',
  ];
  return !textExtensions.includes(ext);
}

class WorkspaceServiceImpl {
  private baseDir: string;
  private workspaces: Map<string, WorkspaceInfo> = new Map();

  constructor() {
    this.baseDir = path.join(fileStorage.getStorageRoot(), 'workspaces');
    this.ensureBaseDir();
  }

  private ensureBaseDir(): void {
    if (!fs.existsSync(this.baseDir)) {
      fs.mkdirSync(this.baseDir, { recursive: true });
    }
  }

  register(workspaceId?: string, sessionId?: string): string {
    const wid = workspaceId || uuidv4().slice(0, 8);
    const sid = sessionId || uuidv4().slice(0, 8);

    if (this.workspaces.has(wid)) {
      const existing = this.workspaces.get(wid)!;
      if (existing.session_id === sid) {
        return wid;
      }
    }

    this.workspaces.set(wid, {
      id: wid,
      session_id: sid,
      status: 'active',
      created_at: new Date().toISOString(),
    });

    const workspacePath = this.getWorkspacePath(sid, wid);
    if (!fs.existsSync(workspacePath)) {
      fs.mkdirSync(workspacePath, { recursive: true });
    }

    return wid;
  }

  getWorkspacePath(sessionId: string, workspaceId: string): string {
    return path.join(this.baseDir, sessionId, workspaceId);
  }

  getWorkspaceInfo(workspaceId: string): WorkspaceInfo | null {
    return this.workspaces.get(workspaceId) || null;
  }

  getWorkspaceBySessionId(sessionId: string): WorkspaceInfo | null {
    for (const info of this.workspaces.values()) {
      if (info.session_id === sessionId) {
        return info;
      }
    }
    return null;
  }

  getWorkspaceDir(workspaceId: string): string | null {
    const info = this.workspaces.get(workspaceId);
    if (!info) return null;
    return this.getWorkspacePath(info.session_id, workspaceId);
  }

  exists(workspaceId: string): boolean {
    return this.workspaces.has(workspaceId);
  }

  listAll(): string[] {
    return Array.from(this.workspaces.keys());
  }

  validatePath(workspaceId: string, targetPath: string): { valid: boolean; path?: string; error?: string } {
    const workspaceDir = this.getWorkspaceDir(workspaceId);
    if (!workspaceDir) {
      return { valid: false, error: `工作区不存在: ${workspaceId}` };
    }

    try {
      const absTarget = path.resolve(targetPath);
      const absWorkspace = path.resolve(workspaceDir);

      if (!absTarget.startsWith(absWorkspace + path.sep) && absTarget !== absWorkspace) {
        return { valid: false, error: `路径越界: ${targetPath} 不在工作区 ${workspaceDir} 范围内` };
      }

      return { valid: true, path: absTarget };
    } catch (err) {
      return { valid: false, error: `路径验证失败: ${err}` };
    }
  }

  isPathAllowed(workspaceId: string, targetPath: string): boolean {
    return this.validatePath(workspaceId, targetPath).valid;
  }

  resolvePath(workspaceId: string, relativePath: string): { valid: boolean; path?: string; error?: string } {
    const workspaceDir = this.getWorkspaceDir(workspaceId);
    if (!workspaceDir) {
      return { valid: false, error: `工作区不存在: ${workspaceId}` };
    }

    if (path.isAbsolute(relativePath)) {
      return this.validatePath(workspaceId, relativePath);
    }

    const fullPath = path.join(workspaceDir, relativePath);
    return this.validatePath(workspaceId, fullPath);
  }

  deleteWorkspace(workspaceId: string): boolean {
    const info = this.workspaces.get(workspaceId);
    if (!info) return false;

    const workspacePath = this.getWorkspacePath(info.session_id, workspaceId);

    try {
      if (fs.existsSync(workspacePath)) {
        fs.rmSync(workspacePath, { recursive: true, force: true });
      }
      this.workspaces.delete(workspaceId);
      return true;
    } catch {
      return false;
    }
  }

  listFiles(workspaceId: string): { success: boolean; files: Array<{ name: string; path: string; is_dir: boolean; size: number; modified_at: string }>; error: string } {
    const workspaceDir = this.getWorkspaceDir(workspaceId);
    if (!workspaceDir) {
      return { success: false, files: [], error: `工作区不存在: ${workspaceId}` };
    }

    if (!fs.existsSync(workspaceDir)) {
      return { success: true, files: [], error: '' };
    }

    const files: Array<{ name: string; path: string; is_dir: boolean; size: number; modified_at: string }> = [];

    try {
      const walk = (dir: string) => {
        const entries = fs.readdirSync(dir, { withFileTypes: true });
        for (const entry of entries) {
          const fullPath = path.join(dir, entry.name);
          const relPath = path.relative(workspaceDir, fullPath).replace(/\\/g, '/');
          const stat = fs.statSync(fullPath);
          if (entry.isDirectory()) {
            files.push({
              name: entry.name,
              path: relPath,
              is_dir: true,
              size: 0,
              modified_at: stat.mtime.toISOString(),
            });
            walk(fullPath);
          } else {
            files.push({
              name: entry.name,
              path: relPath,
              is_dir: false,
              size: stat.size,
              modified_at: stat.mtime.toISOString(),
            });
          }
        }
      };
      walk(workspaceDir);

      files.sort((a, b) => {
        if (a.is_dir !== b.is_dir) return a.is_dir ? -1 : 1;
        return a.path.toLowerCase().localeCompare(b.path.toLowerCase());
      });

      return { success: true, files, error: '' };
    } catch (e) {
      return { success: false, files: [], error: `列出文件失败: ${String(e)}` };
    }
  }

  listSessions(): Record<string, string[]> {
    const sessions: Record<string, string[]> = {};
    for (const [wid, info] of this.workspaces) {
      const sid = info.session_id || 'unknown';
      if (!sessions[sid]) {
        sessions[sid] = [];
      }
      sessions[sid].push(wid);
    }
    return sessions;
  }

  private _getUniqueFilename(dir: string, filename: string): string {
    if (!fs.existsSync(path.join(dir, filename))) {
      return filename;
    }

    const ext = path.extname(filename);
    const base = path.basename(filename, ext);
    let counter = 1;

    while (fs.existsSync(path.join(dir, `${base}_${counter}${ext}`))) {
      counter++;
    }

    return `${base}_${counter}${ext}`;
  }

  async saveUploadedFiles(
    workspaceId: string,
    files: Array<{ filename: string; content: Buffer }>,
    subDir?: string
  ): Promise<{ success: boolean; files: Array<{ original_filename: string; saved_as: string; path: string; size: number }>; error: string }> {
    const workspaceDir = this.getWorkspaceDir(workspaceId);
    if (!workspaceDir) {
      return {
        success: false,
        files: [],
        error: `工作区不存在: ${workspaceId}`,
      };
    }

    let targetDir = workspaceDir;
    if (subDir) {
      const resolved = this.resolvePath(workspaceId, subDir);
      if (!resolved.valid) {
        return {
          success: false,
          files: [],
          error: `无效的子目录路径: ${subDir}`,
        };
      }
      targetDir = resolved.path!;
      if (!fs.existsSync(targetDir)) {
        fs.mkdirSync(targetDir, { recursive: true });
      }
    }

    const savedFiles: Array<{ original_filename: string; saved_as: string; path: string; size: number }> = [];

    try {
      for (const file of files) {
        const uniqueFilename = this._getUniqueFilename(targetDir, file.filename);
        const filePath = path.join(targetDir, uniqueFilename);

        fs.writeFileSync(filePath, file.content);

        savedFiles.push({
          original_filename: file.filename,
          saved_as: uniqueFilename,
          path: filePath,
          size: file.content.length,
        });
      }

      return {
        success: true,
        files: savedFiles,
        error: '',
      };
    } catch (err) {
      return {
        success: false,
        files: savedFiles,
        error: String(err),
      };
    }
  }

  getFileTree(workspaceId: string): { success: boolean; tree?: FileTreeNode; error?: string } {
    const workspaceDir = this.getWorkspaceDir(workspaceId);
    if (!workspaceDir) {
      return { success: false, error: `工作区不存在: ${workspaceId}` };
    }

    if (!fs.existsSync(workspaceDir)) {
      return { success: false, error: `工作区目录不存在` };
    }

    const buildTree = (dir: string, relativePath: string): FileTreeNode | null => {
      const items = fs.readdirSync(dir);
      const filteredItems = items.filter(item => !isHiddenFile(item));

      const children: FileTreeNode[] = [];

      for (const item of filteredItems) {
        const itemPath = path.join(dir, item);
        const itemRelativePath = relativePath ? `${relativePath}/${item}` : item;
        const stats = fs.statSync(itemPath);

        if (stats.isDirectory()) {
          const childNode = buildTree(itemPath, itemRelativePath);
          if (childNode) {
            children.push(childNode);
          }
        } else {
          children.push({
            name: item,
            path: itemRelativePath,
            isDirectory: false,
            size: stats.size,
            type: getMimeType(itemPath),
            modifiedAt: stats.mtime.toISOString(),
          });
        }
      }

      children.sort((a, b) => {
        if (a.isDirectory !== b.isDirectory) {
          return a.isDirectory ? -1 : 1;
        }
        return a.name.localeCompare(b.name);
      });

      return {
        name: path.basename(dir),
        path: relativePath,
        isDirectory: true,
        children,
      };
    };

    const tree = buildTree(workspaceDir, '');
    return { success: true, tree: tree || undefined };
  }

  getFileInfo(workspaceId: string, relativePath: string): { success: boolean; info?: FileInfo; error?: string } {
    const resolved = this.resolvePath(workspaceId, relativePath);
    if (!resolved.valid) {
      return { success: false, error: resolved.error };
    }

    const absolutePath = resolved.path!;
    if (!fs.existsSync(absolutePath)) {
      return { success: false, error: `文件不存在: ${relativePath}` };
    }

    const stats = fs.statSync(absolutePath);
    const workspaceDir = this.getWorkspaceDir(workspaceId)!;
    const normalizedRelativePath = path.relative(workspaceDir, absolutePath).replace(/\\/g, '/');

    return {
      success: true,
      info: {
        name: path.basename(absolutePath),
        path: normalizedRelativePath,
        absolutePath,
        size: stats.isDirectory() ? 0 : stats.size,
        type: stats.isDirectory() ? 'directory' : getMimeType(absolutePath),
        modifiedAt: stats.mtime.toISOString(),
        isDirectory: stats.isDirectory(),
      },
    };
  }

  getFileContent(workspaceId: string, relativePath: string): { success: boolean; content?: string; encoding?: string; size?: number; error?: string } {
    const resolved = this.resolvePath(workspaceId, relativePath);
    if (!resolved.valid) {
      return { success: false, error: resolved.error };
    }

    const absolutePath = resolved.path!;

    if (!fs.existsSync(absolutePath)) {
      return { success: false, error: `文件不存在: ${relativePath}` };
    }

    const stats = fs.statSync(absolutePath);
    if (stats.isDirectory()) {
      return { success: false, error: `不能读取目录内容: ${relativePath}` };
    }

    if (stats.size > MAX_FILE_SIZE) {
      return { success: false, error: `文件过大，超过 1GB 限制: ${relativePath}` };
    }

    const isBinary = isBinaryFile(absolutePath);

    try {
      if (isBinary) {
        const buffer = fs.readFileSync(absolutePath);
        return {
          success: true,
          content: buffer.toString('base64'),
          encoding: 'base64',
          size: stats.size,
        };
      } else {
        const content = fs.readFileSync(absolutePath, 'utf-8');
        return {
          success: true,
          content,
          encoding: 'utf-8',
          size: stats.size,
        };
      }
    } catch (err) {
      return { success: false, error: `读取文件失败: ${err}` };
    }
  }

  deleteFile(workspaceId: string, relativePath: string): { success: boolean; deleted?: boolean; path?: string; error?: string } {
    const resolved = this.resolvePath(workspaceId, relativePath);
    if (!resolved.valid) {
      return { success: false, error: resolved.error };
    }

    const absolutePath = resolved.path!;

    if (!fs.existsSync(absolutePath)) {
      return { success: false, error: `文件不存在: ${relativePath}` };
    }

    try {
      const stats = fs.statSync(absolutePath);
      if (stats.isDirectory()) {
        fs.rmSync(absolutePath, { recursive: true, force: true });
      } else {
        fs.unlinkSync(absolutePath);
      }
      return { success: true, deleted: true, path: relativePath };
    } catch (err) {
      return { success: false, error: `删除失败: ${err}` };
    }
  }
}

export const workspaceService = new WorkspaceServiceImpl();
